### ISSUE TITLE
Support fixed line height for `StyledText`

### ISSUE THREAD

**Issue by SyntevoAlex** [eclipse-platform/eclipse.platform.swt#932]
This is a continuation of https://bugs.eclipse.org/bugs/show_bug.cgi?id=527917

Currently, `StyledText` automatically increases line height when it encounters a character that is taller than current line height.

Here's an example of what happens when it starts with latin text, but then encounters a tall unicode character while scrolling:
![LineHeight17-38](https://github.com/eclipse-platform/eclipse.platform.swt/assets/45797751/6b5f1611-30a6-4dbe-94f5-51fef6dbf67c)

This is bad for two reasons:
1) It wastes a lot of screen space in text editor
2) When two parallel text are needed (for example, to show git diffs), the panes easily desynchronize when scrolling them.

Possible solutions, in order of my preference, include:
1) Somehow render tall characters without increasing line height
2) Have variable line height in `StyledText`
3) Don't render tall unicode characters - that's obviously bad because if character is there, it's there for a reason and should be shown

In my testing, I found that solution 1 is good - "forcing" tall characters into smaller lines still looks reasonable. Here's a screenshot using my not-yet-finished patch on Ubuntu:
![PatchAtWork](https://github.com/eclipse-platform/eclipse.platform.swt/assets/45797751/e466867c-031b-4d12-82a2-c9a7e6a71b7e)

as it can be seen, even the tallest characters still look reasonable without increasing default line height.

**Comment by mickaelistria** [comment]
Code Mining can handle some variable line height and render lines with different height when a character is replaced by a StyleRange with different height. Could this strategy apply to taller unicode character as well, or does it have to be more "native"?

**Comment by SyntevoAlex** [comment]
Sorry but I failed to grasp the idea you're describing.
What is "Code Mining" ? What exactly does it do when a line contains a tall unicode character?

**Comment by SyntevoAlex** [comment]
To clarify, `StyledText` already handles tall characters by making ALL lines tall enough to contain it.

**Comment by mickaelistria** [comment]
Code Minings are what is described in https://eclipse.dev/eclipse/news/4.8/platform_isv.php#codemining-sourceviewer-support . Under the hood, what they do is that they enlarge some lines in the StyledText to display addition information. Enlarging lines is simply achieved by putting the right GlyphMetrics on a StyleRange. Snippet212 does show a simpler case for it, with lines of variable height according to their content.
![Screenshot from 2023-12-05 23-14-33](https://github.com/eclipse-platform/eclipse.platform.swt/assets/469058/4b6c714a-1719-4879-a4c4-18e02b68579c)

**Comment by SyntevoAlex** [comment]
Unfortunately this approach does not help to make lines lower (so that tall characters take same space as latin characters).

I made quick&dirty experimental patches for Windows and Linux.

**Windows**
In `TextLayout.shape()` I made `if (style.metrics != null)` behave as if `metrics` was always set.
The result is that lines now obey requested height. 
However only part of character is rendered - if character is twice taller than line, then only top half of character is painted.
I understand that this produces unreadable text. You can try obscuring bottom half of this text's line and see if it's readable.

**Linux**
In `TextLayout.shape()` I made `if (metrics != null)` behave as if `metrics` was always set.
The result is that line can not become lower, only taller.
Further, `StyledText` does not expect this and still assumes that lines have same height, which throws off hit test calculations.

To summarize, setting `StyleRange` is not a good way to make taller lines have same height as latin lines.

**Comment by revolter** [comment]
> Have variable line height in `StyledText`
I guess that this is more complicated to implement, but does everybody agree that it would be the most correct solution?

**Comment by SyntevoAlex** [comment]
Variable line height is already implemented in `StyledText`.
But when two parallel StyledTexts need to be in sync (to show diffs in git), variable line height becomes a problem of its own.

**Comment by mickaelistria** [comment]
> But when two parallel StyledTexts need to be in sync (to show diffs in git), variable line height becomes a problem of its own.

Can you please clarify what particular problems you're referring to?

**Comment by SyntevoAlex** [comment]
> Can you please clarify what particular problems you're referring to?

1) Aligning (that is, scrolling synchronously) two parallel StyledTexts becomes a bigger problem with variable line height
2) It's uglier. In Linux, gedit uses variable line height, and I never liked it. It's somewhat similar to why we programmers like to use mono fonts.
3) Why waste screen space when hieroglyphs fit rather well into standard latin line heights?

**Comment by mickaelistria** [comment]
> Aligning (that is, scrolling synchronously) two parallel StyledTexts becomes a bigger problem with variable line height

I'm optimistic that this particular issue can be fixed relatively easily. The effort of designing what the behavior should be, then coding it shouldn't be hard.

> Why waste screen space when hieroglyphs fit rather well into standard latin line heights?

Isn't it a bit hard to know in advance whether all characters fit "rather well"?

**Comment by SyntevoAlex** [comment]
> I'm optimistic that this particular issue can be fixed relatively easily. The effort of designing what the behavior should be, then coding it shouldn't be hard.

Maybe. Or maybe not. Either way, it's an additional problem. Also, there's a beauty issue when line of the left has different height from matching line on the right (because one of these has a hieroglyph and the other doesn't). Yes, it's not a very big issue, yet still, this patch solves it.

> Isn't it a bit hard to know in advance whether all characters fit "rather well"?

In the attached testing snippet, I took my time to test all common scripts (Japanese, Chinese, Hebrew, etc etc) on all 3 platforms. It looks good. You don't have to take my word on it, check for yourself.

**Comment by tmssngr** [comment]
See also #3301.
